gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/Text.java

```diff
@@ -29,6 +31,48 @@
  */
 public enum Text implements BiPredicate<String, String> {
 
+    /**
+     * Evaluates if the first string has a regex match with the second (pattern).
+     *
+     * @since 3.6.0
+     */
+    regex {
+        @Override
+        public boolean test(final String value, final String regex) {
+            Pattern pattern = Pattern.compile(regex);
```

Comment:
Pattern compilation is expensive. I am wondering if we need to maintain a LRU cache. In most practical scenarios, it is highly unlikely that user sends a different pattern with every query.

For reference, an benchmark is provided at https://www.baeldung.com/java-regex-pre-compile

Reply:
As I noted in the PR notes at the top, I did consider this and made a conscious decision to defer implementing an LRU (queue/cache) for a later PR. I would like to get the basic functionality in, see how people use it, and then circle back for any LRU type functionality. We use Caffeine in other parts of the project but I'm not sure if that would be too heavy a solution for this.

Reply:
I agree with Divij that Pattern.compile is too costly to do on every solution flowing through the predicate. You don't need an LRU to fix the problem though, just use a new Predicate instance for each regex filter in the query instead of reusing a single Text.regex enum BiPredicate for everything. The RegexPredicate would implement Predicate instead of BiPredicate and take the pattern in the constructor and then just compile it once in the constructor, instead of once per solution at evaluation time. (Obviously this will only work for regex with a constant search pattern, but that should be most of the time.)

Reply:
Actually it would still need to be a BiPredicate because that is what P expect, but at least it would already have the pattern compiled at runtime. The second arg would just be ignored.